File: leetcode/problems/problem-specific/345.reverse-vowels-of-a-string/345.reverse-vowels-of-a-string.go
package main

import "strings"

func reverseVowels(s string) string {
	if len(s) <= 1 {
		return s
	}

	b := []byte(s)
	left, right := 0, len(b)-1
	for left < right {
		if !isVowel(b[left]) && !isVowel(b[right]) {
			left++
			right--
		} else if !isVowel(b[left]) {
			left++
		} else if !isVowel(b[right]) {
			right--
		} else {
			b[left], b[right] = b[right], b[left]
			left++
			right--
		}
	}

	return string(b)
}

func isVowel(b byte) bool {
	res := false
	switch strings.ToLower(string(b)) {
	case "a", "e", "i", "o", "u":
		res = true
	default:
		res = false
	}

	return res
}

File: leetcode/problems/language-specific/python/345.reverse-vowels-of-a-string.py
#
# @lc app=leetcode id=345 lang=python3
#
# [345] Reverse Vowels of a String
#

# @lc code=start

# * Two Pointers Solution | O(n) Time | O(n) Space
# * n -> The length of s string


class Solution:
    def reverseVowels(self, s: str) -> str:
        s_list = list(s)
        left, right = 0, len(s_list) - 1
        while left < right:
            if not self.is_vowel(s_list[left]):
                left += 1
            elif not self.is_vowel(s_list[right]):
                right -= 1
            else:
                s_list[left], s_list[right] = s_list[right], s_list[left]
                left += 1
                right -= 1

        return "".join(s_list)

    def is_vowel(self, char):
        return char.lower() in ("a", "e", "i", "o", "u")


# @lc code=end
